### DISCUSSION TITLE
on_build_file/on_build_files中参数opt的origin方法为nil

### DISCUSSION THREAD

**Discussion by augustheart** [xmake-io/xmake#2534] [Q&A] [answered]
本地xmake版本为xmake v2.6.8+master.39aba1bad
按照文档原样复制的写法，
```lua
target("utilities4")
    set_kind("shared")
    set_targetdir("bin")
    set_prefixname("")
    add_files("utilities/utilities.rc","utilities/src/**.cpp")
    add_includedirs("utilities/include","config","utilities")
    add_links("gdi32")
    on_build_file(function (target, sourcefile, opt)
        opt.origin(target,sourcefile,opt)
    end)
target_end()
```
结果是报错：
```cmd
error: ...mdir\core\sandbox\modules\import\core\base\scheduler.lua:56: field 'origin' is not callable (a nil value)
```

有一个issue提到了这个错误。看issue日志这是18年出来的功能，也没注意到后续的changelog有修改，在线文档也是有这个写法。感觉不应该是bug，求提示
我尝试用  for k,v in pairs(opt) do 打印了opt对象，发现只有一个progress
ps:用这个的原因是想用xmake重写soui的编译脚本，并且尝试使用mingw编译。目前碰到一个rc文件原项目文件是采用utf16保存的，所以我尝试对此文件指定codepage（原则是在非语法问题的前提下尽量不修改原项目文件）

**Comment by waruqi** [comment, marked as answer]
opt.origin 这是早期遗留设计，应该已经不存在了，文档部分有些可能没更新。。你可以在 before_build_file/after_build_file 里面操作
